### PR TITLE
PXC-3760: Implement a feature to evict a lagging node from PXC

### PR DESCRIPTION
**Pull request by voodam**
https://jira.percona.com/browse/PXC-3760

Problem:
We want to evict a node if it slows down a whole cluster due to the flow control.

Solution:
We introduce two variables: gcs.fc_auto_evict_window and gcs.fc_auto_evict_threshold. If the node pauses in FC more then "threshold * window" in the last "window" seconds frame, node will just abort.

The abortion can be problematic if we have 2-nodes cluster. From the side of other nodes this will be indistinguishable from the crash of the lagging node.

### CODE REVIEW

File: mysql-test/suite/galera/r/galera_restart_on_unknown_option.result [hunk 1]
```diff
@@ -4,6 +4,8 @@ call mtr.add_suppression("The privilege system failed to initialize correctly");
 call mtr.add_suppression("SST failed/interrupted");
 call mtr.add_suppression("IST was interrupted");
 call mtr.add_suppression("IST didn't contain all write sets");
+call mtr.add_suppression("Attempting backtrace");
```
Comment (kamil-holubicki) [inline mysql-test/suite/galera/r/galera_restart_on_unknown_option.result]: I think this one was fixed on 8.0.32 release branch.
@venkatesh-prasad-v

File: mysql-test/suite/galera/t/galera_fc_auto_evict.test [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+--connection node_2
+--source include/expect_crash.inc
+# Triggering flow control logic taken from here:
+# https://www.percona.com/blog/monitoring-flow-control-percona-xtradb-cluster-5-6/
+SET GLOBAL wsrep_provider_options="gcs.fc_limit=1;gcs.fc_auto_evict_window=0.1";
+FLUSH TABLES WITH READ LOCK;
+
+--connection node_1
+CREATE TABLE t1(a INT);
```
Comment (kamil-holubicki) [inline mysql-test/suite/galera/t/galera_fc_auto_evict.test]: Doesn't it complain about the lack of the explicit primary key?

Comment (kamil-holubicki) [inline mysql-test/suite/galera/t/galera_fc_auto_evict.test]: Please wrap inserting loop in `--disable_query_log` / `--enable_query_log` and add `--echo ...` informing about what is happening.

**Comment by kamil-holubicki** [comment]
https://pxc.cd.percona.com/job/pxc-8.0-pipeline-parallel-mtr/1613/

File: mysql-test/suite/galera/t/galera_fc_auto_evict.test (hunk 2 above)
Comment (voodam) [reply mysql-test/suite/galera/t/galera_fc_auto_evict.test]: No, where this logic is located? Or the name of error. I couldn't find anything quickly.

File: mysql-test/suite/galera/r/galera_restart_on_unknown_option.result (hunk 1 above)
Comment (voodam) [reply mysql-test/suite/galera/r/galera_restart_on_unknown_option.result]: Revert this?

File: mysql-test/suite/galera/t/galera_fc_auto_evict.test (hunk 2 above)
Comment (voodam) [reply mysql-test/suite/galera/t/galera_fc_auto_evict.test]: Done

File: mysql-test/suite/galera/r/galera_restart_on_unknown_option.result (hunk 1 above)
Comment (venkatesh-prasad-v) [reply mysql-test/suite/galera/r/galera_restart_on_unknown_option.result]: Yes. This has been fixed by commit https://github.com/percona/percona-xtradb-cluster/commit/c01b796d8fc31cb8352817fb7e79bb22646db67c. So, please revert this change.

File: mysql-test/suite/galera/t/galera_fc_auto_evict.test (hunk 2 above)
Comment (venkatesh-prasad-v) [reply mysql-test/suite/galera/t/galera_fc_auto_evict.test]: PXC allows creation of tables without primary keys, however it [complains](https://github.com/venkatesh-prasad-v/percona-xtradb-cluster/blob/8.0/sql/sql_base.cc#L6219-L6224) when we try to execute a DMLs  when pxc_strict_mode=ENFORCING.

File: mysql-test/suite/galera/t/galera_fc_auto_evict.test [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+--connection node_2
+UNLOCK TABLES;
+
+--connection node_1
+--let $wait_condition=SELECT VARIABLE_VALUE = 1 FROM performance_schema.global_status WHERE VARIABLE_NAME = 'wsrep_cluster_size'
+--source include/wait_condition.inc
+
+--let SEARCH_FILE=$MYSQLTEST_VARDIR/log/mysqld.2.err
+--let SEARCH_PATTERN=FC auto evict: pause limit exceeded
+--source include/search_pattern.inc
```
Comment (venkatesh-prasad-v) [inline mysql-test/suite/galera/t/galera_fc_auto_evict.test]: Please consider using `assert_grep.inc` as it fails immediately when the required pattern is not found.

File: mysql-test/suite/galera/t/galera_fc_auto_evict.test [hunk 4]
```diff
@@ -0,0 +1,54 @@
+# PXC-3760: Implement a feature to evict a lagging node from PXC
+#
+# Problem:
+# We want to evict a node if it slows down a whole cluster due to the flow control.
+#
+# Solution:
+# We introduce two variables: gcs.fc_auto_evict_window and gcs.fc_auto_evict_threshold.
+# If the node pauses in FC more then "threshold * window" in the last  window" seconds
```
Comment (venkatesh-prasad-v) [inline mysql-test/suite/galera/t/galera_fc_auto_evict.test]:
```suggestion
# If the node pauses in FC more than "threshold * window" in the last "window" seconds
```

File: mysql-test/suite/galera/r/galera_defaults.result [hunk 5]
```diff
@@ -24,7 +25,6 @@ WSREP_CERTIFICATION_RULES	strict
 WSREP_CERTIFY_NONPK	ON
 WSREP_CLUSTER_NAME	my_wsrep_cluster
 WSREP_DBUG_OPTION	
-WSREP_DEBUG	NONE
```
Comment (venkatesh-prasad-v) [inline mysql-test/suite/galera/r/galera_defaults.result]: Why this change?

Comment (voodam) [reply mysql-test/suite/galera/r/galera_defaults.result]: Probably it isn't required, I'll revert this too. At first I thought this param depends on build type, not on CLI parameter.

File: mysql-test/suite/galera/t/galera_fc_auto_evict.test (hunk 4 above)
Comment (voodam) [reply mysql-test/suite/galera/t/galera_fc_auto_evict.test]: Done

File: mysql-test/suite/galera/t/galera_fc_auto_evict.test (hunk 2 above)
Comment (kamil-holubicki) [reply mysql-test/suite/galera/t/galera_fc_auto_evict.test]:
```
mysql> create table t1 (a int);
Query OK, 0 rows affected (0,19 sec)

mysql> insert into t1 values (0);
ERROR 1105 (HY000): Percona-XtraDB-Cluster prohibits use of DML command on a table (test.t1) without an explicit primary key with pxc_strict_mode = ENFORCING or MASTER
```
Aren't we in `ENFORCING` mode during the test?

Comment (voodam) [reply mysql-test/suite/galera/t/galera_fc_auto_evict.test]: In my case it's `DISABLED`

Comment (venkatesh-prasad-v) [reply mysql-test/suite/galera/t/galera_fc_auto_evict.test]: It has been set to DISABLED by default for all test cases. See https://github.com/venkatesh-prasad-v/percona-xtradb-cluster/blob/8.0/mysql-test/include/default_mysqld.cnf#L27

File: mysql-test/suite/galera/t/galera_fc_auto_evict.test (hunk 3 above)
Comment (voodam) [reply mysql-test/suite/galera/t/galera_fc_auto_evict.test]: Done

File: mysql-test/suite/galera/t/galera_fc_auto_evict.test (hunk 2 above)
Comment (voodam) [reply mysql-test/suite/galera/t/galera_fc_auto_evict.test]: Anyway, rewrited with primary key

File: mysql-test/suite/galera/t/galera_fc_auto_evict.test [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+--connection node_2
+UNLOCK TABLES;
+
+--connection node_1
+--let $wait_condition=SELECT VARIABLE_VALUE = 1 FROM performance_schema.global_status WHERE VARIABLE_NAME = 'wsrep_cluster_size'
+--source include/wait_condition.inc
+
+--let $assert_text= FС auto eviction triggered
+--let $assert_file= $MYSQLTEST_VARDIR/log/mysqld.2.err
+--let $assert_select= FC auto evict: pause limit exceeded
```
Comment (venkatesh-prasad-v) [inline mysql-test/suite/galera/t/galera_fc_auto_evict.test]: Please update the test since the error message was changed.

**Comment by kamil-holubicki** [comment]
Please solve the conflict after merging galera

**Review by kamil-holubicki — APPROVED** [review]

File: mysql-test/suite/galera/t/galera_fc_auto_evict.test (hunk 6 above)
Comment (voodam) [reply mysql-test/suite/galera/t/galera_fc_auto_evict.test]: Yes, sure

**Review by venkatesh-prasad-v — APPROVED** [review]

**Comment by kamil-holubicki** [comment]
Looks like the new test fails on Jenkins
https://pxc.cd.percona.com/job/pxc-8.0-pipeline-parallel-mtr/1617/

**Comment by voodam** [comment]
https://pxc.cd.percona.com/job/pxc-8.0-pipeline-parallel-mtr/1618/

**Review by kamil-holubicki — APPROVED** [review]
